### PR TITLE
SquashDictInnerAssertLenKeys

### LINKED ISSUE #302: SquashDictInnerAssertLenKeys
Asserts the length of the keys array during the squashing process.

Reference: https://github.com/starkware-libs/cairo-lang/blob/efa9648f57568aad8f8a13fbf027d2de7c63c2c0/src/starkware/cairo/common/squash_dict.cairo#L216

depends on #245

### PR DESCRIPTION
**Pull request by jbmkahdeksan**
resolves #302 

Added Squash Dict Inner Assert Len Keys

### CODE REVIEW

**Review by rodrigo-pino — COMMENTED** [review]
Great! Could you add a unit test for this hint?

**Review by TAdev0 — CHANGES_REQUESTED** [review]
@jbmkahdeksan you have a linting issue to resolve.

otherwise lgtm!

File: pkg/hintrunner/zero/zerohint.go [hunk 1]
```diff
@@ -137,6 +137,9 @@ func GetHintFromCode(program *zero.ZeroProgram, rawHint zero.Hint, hintPC uint64
 		// Usort hints
 	case usortEnterScopeCode:
 		return createUsortEnterScopeHinter()
+		// Dicitonaries hints
```
Comment (TAdev0) [inline pkg/hintrunner/zero/zerohint.go:140]: Dicitonaries -> Dictionaries

File: pkg/hintrunner/zero/zerohint_dictionaries.go [hunk 2]
```diff
@@ -0,0 +1,24 @@
+package zero
+
+import (
+	"fmt"
+	"github.com/NethermindEth/cairo-vm-go/pkg/hintrunner/hinter"
+	VM "github.com/NethermindEth/cairo-vm-go/pkg/vm"
+)
+
```
[the remark this answers is not in the document]
Comment (jbmkahdeksan) [reply pkg/hintrunner/zero/zerohint_dictionaries.go:8]: didn't notice this, thank you! 👍

File: pkg/hintrunner/zero/zerohint.go (hunk 1 above)
Comment (jbmkahdeksan) [reply pkg/hintrunner/zero/zerohint.go:140]: Oops! 😅

File: pkg/hintrunner/zero/zerohint_dictionaries_test.go [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+	"testing"
+
+	"github.com/NethermindEth/cairo-vm-go/pkg/hintrunner/hinter"
+	"github.com/consensys/gnark-crypto/ecc/stark-curve/fp"
+)
+
+func TestZeroHintDictionaries(t *testing.T) {
+	runHinterTests(t, map[string][]hintTestCase{
+		"SquashDictInnerAssertKeys": {
+			{
```
Comment (TAdev0) [inline pkg/hintrunner/zero/zerohint_dictionaries_test.go:13]: I think you can have only one test : 

```
			{
				operanders: []*hintOperander{},
				ctxInit: func(ctx *hinter.HintRunnerContext) {
					err := hinter.InitializeSquashedDictionaryManager(ctx)
					if err != nil {
						t.Fatal(err.Error())
					}
				},
				makeHinter: func(ctx *hintTestContext) hinter.Hinter {
					return newSquashDictInnerAssertLenKeysHint()
				},
				check: func(t *testing.T, ctx *hintTestContext) {
				},
			},
```

actually, you dont need anything in `check`, because the check is the hint itself.

The `ctxInit` step will call `InitializeSquashedDictionaryManager`  which creates a `SquashedDictionaryManager` without any key in the list.

Then the hint will do the check that `len == 0`

If you try this : 

```
		"SquashDictInnerAssertKeys": {
			{
				operanders: []*hintOperander{},
				ctxInit: func(ctx *hinter.HintRunnerContext) {
					err := hinter.InitializeSquashedDictionaryManager(ctx)
					if err != nil {
						t.Fatal(err.Error())
					}

					ctx.SquashedDictionaryManager.Keys = append(ctx.SquashedDictionaryManager.Keys, fp.Element{})

				},
				makeHinter: func(ctx *hintTestContext) hinter.Hinter {
					return newSquashDictInnerAssertLenKeysHint()
				},
				check: func(t *testing.T, ctx *hintTestContext) {
				},
			},
```

The test will fail because of `ctx.SquashedDictionaryManager.Keys = append(ctx.SquashedDictionaryManager.Keys, fp.Element{})` during the `ctxInit` , and its actually the code of the hint that will fail.

Comment (TAdev0) [reply pkg/hintrunner/zero/zerohint_dictionaries_test.go:13]: The test consists more of making the initial setup and running the hint than making a check after the hint

**Review by TAdev0 — CHANGES_REQUESTED** [review]

File: pkg/hintrunner/zero/zerohint_dictionaries_test.go (hunk 3 above)
Comment (TAdev0) [reply pkg/hintrunner/zero/zerohint_dictionaries_test.go:13]: if there is a way in go to indicate that test should actually have failing code to pass, then you can add a second test where you add some key in the list before calling the hint !

Comment (jbmkahdeksan) [reply pkg/hintrunner/zero/zerohint_dictionaries_test.go:13]: Got it, thanks a lot!! ✌️
I was actually a little confused with this part and learned a lot from this review!

Comment (jbmkahdeksan) [reply pkg/hintrunner/zero/zerohint_dictionaries_test.go:12]: new changes were pushed!
thanks a lot @TAdev0 🫡

**Review by TAdev0 — APPROVED** [review]

**Comment by TAdev0** [comment]
LGTM ! @rodrigo-pino  @chee-chyuan  @MaksymMalicki  ready for final review before merging

**Comment by har777** [comment]
I'm not really sure about this.

`assert len(keys) == 0` seems to assert that the length of value assigned to scope value `keys` is zero. Don't think directly checking the length of the squashed dict is the same 🤔 IMO you should just fetch the scope value of `keys` as a list and then assert its length is 0. @cicr99 

reference -> `keys` is being set [here](https://github.com/starkware-libs/cairo-lang/blob/efa9648f57568aad8f8a13fbf027d2de7c63c2c0/src/starkware/cairo/common/squash_dict.cairo#L55)

Also another test where the assertion fails would be ideal.

**Comment by TAdev0** [comment]
@har777  i just checked , indeed it seems that `keys` should be retrieve as a scope variable. but i think that checking the descending list `Keys` which is a field of the `SquashedDictionaryManager` (manager of all squashed dicts) should work as well?  But yeah i think you're right and we should fetch keys in the same way as the python hint

File: pkg/hintrunner/core/hint.go [hunk 4]
```diff
@@ -978,6 +978,11 @@ func (hint *InitSquashData) Execute(vm *VM.VirtualMachine, ctx *hinter.HintRunne
 		return fmt.Errorf("get first key: %w", err)
 	}
 
+	err = ctx.ScopeManager.AssignVariable("keys", ctx.SquashedDictionaryManager.Keys)
+	if err != nil {
+		return err
+	}
```
Comment (har777) [inline pkg/hintrunner/core/hint.go:984]: Why is this required? I was thinking some other hint will be populating the value of `keys`. We don't need to do anything extra. cc @cicr99

Comment (jbmkahdeksan) [reply pkg/hintrunner/core/hint.go:984]: I don't understand what you mean. Could you please elaborate? Maybe via Telegram...

Comment (TAdev0) [reply pkg/hintrunner/core/hint.go:984]: why would you need to modify the _hint.go_ file?

Comment (jbmkahdeksan) [reply pkg/hintrunner/core/hint.go:984]: I had no idea before, I thought 'keys' had to be obtain it's somewhere in order for this hint to work...

Comment (jbmkahdeksan) [reply pkg/hintrunner/core/hint.go]: new changes pushed @TAdev0 @har777 @rodrigo-pino @cicr99

**Comment by har777** [comment]
lgtm!

File: pkg/hintrunner/zero/zerohint_dictionaries_test.go [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+				operanders: []*hintOperander{},
+				ctxInit: func(ctx *hinter.HintRunnerContext) {
+					ctx.ScopeManager.EnterScope(map[string]any{"keys": []fp.Element{}})
+					err := hinter.InitializeSquashedDictionaryManager(ctx)
```
Comment (TAdev0) [inline pkg/hintrunner/zero/zerohint_dictionaries_test.go:17]: Why is it needed to call `InitializeSquashedDictionaryManager` here? 
If we purely test the hint, then we just need to enter the scope with keys.

If you remove it , test pass. `InitializeSquashedDictionaryManager`   deals with the `SquashedDictionaryManager` which shouldn't be managed manually for this hint anyway as previously discussed.

You dont use it in the second test, i guess it can be removed ? @har777  @cicr99

**Review by TAdev0 — CHANGES_REQUESTED** [review]

File: pkg/hintrunner/zero/zerohint_dictionaries_test.go (hunk 5 above)
Comment (har777) [reply pkg/hintrunner/zero/zerohint_dictionaries_test.go:17]: Yep just `hinter.InitializeScopeManager(ctx, map[string]any{"keys": []fp.Element{}})` is enough

Comment (jbmkahdeksan) [reply pkg/hintrunner/zero/zerohint_dictionaries_test.go:17]: thank you, will fix it!

Comment (jbmkahdeksan) [reply pkg/hintrunner/zero/zerohint_dictionaries_test.go]: sure, you are right!
`InitializeSquashedDictionaryManager` is beyond the hint implementation
I don't need it to test the hint because the hint itself is a granular and independent part from the DictionaryManager and I only need a scope to test the hint.
Is that correct?

Comment (jbmkahdeksan) [reply pkg/hintrunner/zero/zerohint_dictionaries_test.go]: @TAdev0 @har777 new changes were pushed! 
thank you, I value a lot your feedback!

**Review by TAdev0 — APPROVED** [review]

**Review by cicr99 — APPROVED** [review]
